perception/sandbox/polygrasp/src/polygrasp/robot_interface.py

```diff
@@ -0,0 +1,245 @@
+
+
+def grasp_to_pose(grasp: graspnetAPI.Grasp):
+    return grasp.translation, R.from_matrix(grasp.rotation_matrix).as_quat()
+
+
+def min_dist_grasp(default_ee_quat, grasps):
+    """Find the grasp with minimum orientation distance to the reference grasp"""
+    with torch.no_grad():
+        rots_as_quat = [
+            torch.Tensor(R.from_matrix(grasp.rotation_matrix).as_quat())
+            for grasp in grasps
+        ]
+        dists = [compute_quat_dist(rot, default_ee_quat) for rot in rots_as_quat]
+        i = torch.argmin(torch.Tensor(dists)).item()
+    log.info(f"Grasp {i} has angle {dists[i]} from reference.")
+    return grasps[i], i
+
+
+def min_dist_grasp_no_z(default_ee_quat, grasps):
+    """
+    Find the grasp with minimum orientation distance to the reference grasp
+    disregarding orientation about z axis
+    """
+    with torch.no_grad():
+        rots_as_R = [R.from_quat(compute_des_pose(grasp)[2]) for grasp in grasps]
+        default_r = R.from_quat(default_ee_quat)
+        dists = [
+            np.linalg.norm((rot.inv() * default_r).as_rotvec()[:2]) for rot in rots_as_R
+        ]
+        i = torch.argmin(torch.Tensor(dists)).item()
+    log.info(f"Grasp {i} has angle {dists[i]} from reference.")
+    return grasps[i], i
+
+
+class GraspingRobotInterface(polymetis.RobotInterface):
+    def __init__(
+        self,
+        gripper: polymetis.GripperInterface,
+        k_approach=1.5,
+        k_grasp=0.72,
+        gripper_max_width=0.085,
+        # ikpy params:
+        base_elements=["panda_link0"],
+        soft_limits=[
+            (-2.70, 2.70),
+            (-1.56, 1.56),
+            (-2.7, 2.7),
+            (-2.87, -0.07),
+            (-2.7, 2.7),
+            (-0.02, 3.55),
+            (-2.7, 2.7),
+        ],
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.gripper = hydra.utils.instantiate(gripper)
+
+        self.default_ee_quat = torch.Tensor([1, 0, 0, 0])
+        self.k_approach = k_approach
+        self.k_grasp = k_grasp
+        self.gripper_max_width = gripper_max_width
+
+        with tempfile.NamedTemporaryFile(mode="w+") as f:
+            f.write(self.metadata.urdf_file)
+            f.seek(0)
+            self.robot_model_ikpy = ikpy.chain.Chain.from_urdf_file(
+                f.name,
+                base_elements=base_elements,
+            )
+        for i in range(len(soft_limits)):
+            self.robot_model_ikpy.links[i + 1].bounds = soft_limits[i]
+
+    def ik(self, position, orientation=None):
+        curr_joint_pos = [0] + self.get_joint_positions().numpy().tolist() + [0]
+        des_homog_transform = np.eye(4)
+        if orientation is not None:
+            des_homog_transform[:3, :3] = R.from_quat(orientation).as_matrix()
+        des_homog_transform[:3, 3] = position
+        try:
+            joint_pos_ikpy = self.robot_model_ikpy.inverse_kinematics_frame(
```

Comment:
Just to confirm, the IK implementation in ikpy is deterministic, correct? If there are sampling-based steps, you might deem a grasp valid when performing ik to check feasibility, but unable to achieve the same joint pos answer when you try to move to it.

Reply:
Yeah, it's just a `scipy.optimize.least_squares` optimization from a fixed init